### PR TITLE
POC: template variables support for public dashboards

### PR DESCRIPTION
**Pull request by evictorero**
This is just a simple POC for template variables support on Public Dashboards, not for merge.

### CODE REVIEW

File: pkg/services/publicdashboards/service/service.go [hunk 1]
```diff
@@ -253,8 +256,48 @@ func (pd *PublicDashboardServiceImpl) buildMetricRequest(ctx context.Context, da
...
+
+			name := variable.Get("name").MustString()
+
+			// some DS uses rawSQL
+			rawSql := queries[i].Get("rawSql").MustString()
+			rawSql = strings.Replace(rawSql, fmt.Sprintf("${%s}", name), value, -1)
+			rawSql = strings.Replace(rawSql, fmt.Sprintf("$%s", name), value, -1)
+			queries[i].Set("rawSql", rawSql)
+
+			// some DS uses expr
```
Comment (timlevett) [inline pkg/services/publicdashboards/service/service.go:295]: there is so many "some DS uses x" scenarios, i wonder if there is a generic approach. Could we string replace the whole blob?

Comment (torkelo) [reply pkg/services/publicdashboards/service/service.go:295]: you need to send the variables (name+value) to the data source in the data source request query model, each data source will need to implement the interpolation

File: pkg/services/publicdashboards/service/service.go [hunk 2]
```diff
@@ -253,8 +256,48 @@ func (pd *PublicDashboardServiceImpl) buildMetricRequest(ctx context.Context, da
...
+			} else {
+				value = variable.Get("current").Get("value").MustString()
+			}
+
+			name := variable.Get("name").MustString()
+
+			// some DS uses rawSQL
+			rawSql := queries[i].Get("rawSql").MustString()
+			rawSql = strings.Replace(rawSql, fmt.Sprintf("${%s}", name), value, -1)
+			rawSql = strings.Replace(rawSql, fmt.Sprintf("$%s", name), value, -1)
```
Comment (torkelo) [inline pkg/services/publicdashboards/service/service.go:292]: We have some more complex syntaxes that support formatting option as well:
https://grafana.com/docs/grafana/latest/dashboards/variables/variable-syntax/ 

```${var_name:<format>}```

here are the formats: https://github.com/grafana/grafana/blob/main/public/app/features/templating/formatRegistry.ts 

We need a separate TemplateInterpolationService I think which each data source can use to interpolate the query. But I guess that is not going to work well for external backend data sources. Maybe each data source can in plugin.json define how to interpolate
strings by default? tricky stuff.

File: pkg/services/publicdashboards/service/service.go (hunk 1 above)
Comment (torkelo) [reply pkg/services/publicdashboards/service/service.go:295]: but that is going to be hard for external data sources unless move the interpolation code and variable formats into the go sdk

**Pull request closed without merging**